### PR TITLE
qt-5: add qtWrapAllExecutables flag, default for Darwin

### LINKED ISSUE #102044: wrapQtAppsHook broken on macOS
**Describe the bug**

The convenience function `wrapQtAppsHook` does not work on macOS. The calls `patchelf` like so:
```
            patchelf --print-interpreter "$file" >/dev/null 2>&1 || continue
```
If the calls succeeds `wrapQtApp` is called with`$file` as argument. Since `patchelf` is not installed, this always fails, skip all files, and binaries are not properly wrapped – and hence fail to execute.

As a work-around I had to call `wrapQtApp` manually in the `preFixup` phase. It did it's thing, wrapped the binary properly and I could run the binary without issues.

**To Reproduce**
1. Build an arbitrary Qt app on macOS (for example gpxlab or glogg, but really any Qt app works …)
2. Compile and install under Linux, to see wrapping working
3. Disable manual call to `wrapQtApp`
4. Compile and install under macOS to see wrapping via the hook doing nothing
5. Verify that the app under macOS is indeed not wrapped and fails to execute

**Expected behavior**

Properly wrapped application. Wrapping itself works, as evidenced by a manual call to `wrapQtApp`, but have the hook not depend on `patchelf` on macOS or at least ** give a warning **  and give a warning in the manual.

**Screenshots**

**Additional context**

**Notify maintainers**

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.01 MiB download, 0.03 MiB unpacked):
  /nix/store/12pkkc7l378r75xz1kvh39wlk1p4gw7w-DarwinTools-1
  /nix/store/zvir75d0b5swjva0xkyxcidpagf7jpdv-nix-info
copying path '/nix/store/12pkkc7l378r75xz1kvh39wlk1p4gw7w-DarwinTools-1' from 'https://cache.nixos.org'...
copying path '/nix/store/zvir75d0b5swjva0xkyxcidpagf7jpdv-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.6.0, macOS 10.15.7`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.7`
 - channels(hannesweisbach): `""`
 - channels(root): `"nixpkgs-20.09pre242465.bd649f4f55c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```
Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by graham33**
###### Motivation for this change

This wraps everything executable rather than just ELF binaries.  Note as a
side-effect, non-executable ELF files will no longer be wrapped on any
platform.  This is to ensure Darwin (Mach-O) executables are wrapped.

There might be a better approach, but this at least allows pkgs.octaveFull
to be wrapped correctly.  This should fix #102044.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131378` at 9462675f run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages marked as broken and skipped:</summary>
  <ul>
    <li>colmapWithCuda</li>
    <li>easyeffects</li>
    <li>invoice2data</li>
    <li>keyfinder</li>
    <li>mame</li>
    <li>obs-studio-plugins.obs-ndi</li>
    <li>openems</li>
    <li>pdfdiff</li>
    <li>pulseeffects-legacy</li>
    <li>python38Packages.foxdot</li>
    <li>python38Packages.python-csxcad</li>
    <li>python38Packages.python-openems</li>
    <li>python39Packages.foxdot</li>
    <li>python39Packages.python-csxcad</li>
    <li>python39Packages.python-openems</li>
    <li>qlandkartegt</li>
    <li>sonic-pi</li>
    <li>tvheadend</li>
    <li>twinkle</li>
    <li>unicap</li>
    <li>xpdf</li>
  </ul>
</details>
<details>
  <summary>22 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/7cc972a57ddc4a14552a5cc7c6e5f0b6">libsForQt512.karchive</a></li>
    <li><a href="https://gist.github.com/33621a2548d9c14e9de9d80546f67103">libsForQt512.kauth</a></li>
    <li><a href="https://gist.github.com/4028407045532955be02beddb8d4a053">libsForQt512.kcodecs</a></li>
    <li><a href="https://gist.github.com/3430937e7b0435de1f04d4f83812eda2">libsForQt512.kcompletion</a></li>
    <li><a href="https://gist.github.com/0e2faa7248efe2e0a8e4e7deafd88d8b">libsForQt512.kconfig</a></li>
    <li><a href="https://gist.github.com/ac4c2098717eccfef53df9eef1be11c0">libsForQt512.kconfigwidgets</a></li>
    <li><a href="https://gist.github.com/7bc00b701b71fd720a79f9f283482838">libsForQt512.kcoreaddons</a></li>
    <li><a href="https://gist.github.com/a9eeb249b6151167f8ed91cb45a20fe7">libsForQt512.kcrash</a></li>
    <li><a href="https://gist.github.com/b1609fca1b8b1f349c5f8e996ac9bc11">libsForQt512.kdbusaddons</a></li>
    <li><a href="https://gist.github.com/585ed66590ba7d718faea56015daad58">libsForQt512.kdoctools</a></li>
    <li><a href="https://gist.github.com/9a1c3aa089929460b6fffd0729d439cd">libsForQt512.kglobalaccel</a></li>
    <li><a href="https://gist.github.com/d8ba380e0d221eb103a7701df17763eb">libsForQt512.kguiaddons</a></li>
    <li><a href="https://gist.github.com/6906a4448c92f6101639b725e2fc213e">libsForQt512.ki18n</a></li>
    <li><a href="https://gist.github.com/ebcbf4ae50d676e0bcf85e59db7534ce">libsForQt512.kiconthemes</a></li>
    <li><a href="https://gist.github.com/be13b7d9f79aadd2b84f591703a7dcc3">libsForQt512.kitemviews</a></li>
    <li><a href="https://gist.github.com/28e7847c339182c669c59b476478ee75">libsForQt512.kjobwidgets</a></li>
    <li><a href="https://gist.github.com/e384214b7d043faf7fe500ecee3e0380">libsForQt512.kpackage</a></li>
    <li><a href="https://gist.github.com/777325b5c3d49aa739f282281d317822">libsForQt512.kservice</a></li>
    <li><a href="https://gist.github.com/e84fd558cf34455d5220295b8f4a91b8">libsForQt512.ktextwidgets</a></li>
    <li><a href="https://gist.github.com/c3af0d59b351dc544310dc0b99fbc223">libsForQt512.kwidgetsaddons</a></li>
    <li><a href="https://gist.github.com/5807aa0ea67a03a24f52096a3970b015">libsForQt512.kwindowsystem</a></li>
    <li><a href="https://gist.github.com/e5176b778598b8f255599d456210dc34">libsForQt512.sonnet</a></li>
  </ul>
</details>
<details>
  <summary>1481 packages skipped due to time constraints:</summary>
  <ul>
    <li>AusweisApp2</li>
    <li>adwaita-qt</li>
    <li>aj-snapshot</li>
    <li>ajour</li>
    <li>albert</li>
    <li>ams</li>
    <li>android-file-transfer</li>
    <li>anki</li>
    <li>ankisyncd</li>
    <li>antimicroX</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>68 packages built successfully:</summary>
  <ul>
    <li>extra-cmake-modules (libsForQt5.extra-cmake-modules ,libsForQt515.extra-cmake-modules ,plasma5Packages.extra-cmake-modules)</li>
    <li>libsForQt5.breeze-icons (libsForQt515.breeze-icons ,plasma5Packages.breeze-icons)</li>
    <li>libsForQt5.karchive (libsForQt515.karchive ,plasma5Packages.karchive)</li>
    <li>libsForQt5.kauth (libsForQt515.kauth ,plasma5Packages.kauth)</li>
    <li>libsForQt5.kcodecs (libsForQt515.kcodecs ,plasma5Packages.kcodecs)</li>
    <li>libsForQt5.kcolorpicker (libsForQt515.kcolorpicker ,plasma5Packages.kcolorpicker)</li>
    <li>libsForQt5.kcompletion (libsForQt515.kcompletion ,plasma5Packages.kcompletion)</li>
    <li>libsForQt5.kconfig (libsForQt515.kconfig ,plasma5Packages.kconfig)</li>
    <li>libsForQt5.kconfigwidgets (libsForQt515.kconfigwidgets ,plasma5Packages.kconfigwidgets)</li>
    <li>libsForQt5.kcoreaddons (libsForQt515.kcoreaddons ,plasma5Packages.kcoreaddons)</li>
    <li>libsForQt5.kcrash (libsForQt515.kcrash ,plasma5Packages.kcrash)</li>
    <li>libsForQt5.kdbusaddons (libsForQt515.kdbusaddons ,plasma5Packages.kdbusaddons)</li>
    <li>libsForQt5.kdoctools (libsForQt515.kdoctools ,plasma5Packages.kdoctools)</li>
    <li>libsForQt5.kglobalaccel (libsForQt515.kglobalaccel ,plasma5Packages.kglobalaccel)</li>
    <li>libsForQt5.kguiaddons (libsForQt515.kguiaddons ,plasma5Packages.kguiaddons)</li>
    <li>libsForQt5.ki18n (libsForQt515.ki18n ,plasma5Packages.ki18n)</li>
    <li>libsForQt5.kiconthemes (libsForQt515.kiconthemes ,plasma5Packages.kiconthemes)</li>
    <li>libsForQt5.kimageannotator (libsForQt515.kimageannotator ,plasma5Packages.kimageannotator)</li>
    <li>libsForQt5.kitemviews (libsForQt515.kitemviews ,plasma5Packages.kitemviews)</li>
    <li>libsForQt5.kjobwidgets (libsForQt515.kjobwidgets ,plasma5Packages.kjobwidgets)</li>
    <li>libsForQt5.knotifications (libsForQt515.knotifications ,plasma5Packages.knotifications)</li>
    <li>libsForQt5.kpackage (libsForQt515.kpackage ,plasma5Packages.kpackage)</li>
    <li>libsForQt5.kplotting (libsForQt515.kplotting ,plasma5Packages.kplotting)</li>
    <li>libsForQt5.kservice (libsForQt515.kservice ,plasma5Packages.kservice)</li>
    <li>libsForQt5.ktextwidgets (libsForQt515.ktextwidgets ,plasma5Packages.ktextwidgets)</li>
    <li>libsForQt5.kwallet (libsForQt515.kwallet ,plasma5Packages.kwallet)</li>
    <li>libsForQt5.kwidgetsaddons (libsForQt515.kwidgetsaddons ,plasma5Packages.kwidgetsaddons)</li>
    <li>libsForQt5.kwindowsystem (libsForQt515.kwindowsystem ,plasma5Packages.kwindowsystem)</li>
    <li>libsForQt5.phonon (libsForQt515.phonon ,plasma5Packages.phonon)</li>
    <li>libsForQt5.plasma-wayland-protocols (libsForQt515.plasma-wayland-protocols ,plasma5Packages.plasma-wayland-protocols)</li>
    <li>libsForQt5.polkit-qt (libsForQt515.polkit-qt ,plasma5Packages.polkit-qt)</li>
    <li>libsForQt5.sonnet (libsForQt515.sonnet ,plasma5Packages.sonnet)</li>
    <li>libsForQt512.breeze-icons</li>
    <li>libsForQt512.extra-cmake-modules</li>
    <li>libsForQt512.kcolorpicker</li>
    <li>libsForQt512.kimageannotator</li>
    <li>libsForQt512.plasma-wayland-protocols</li>
    <li>libsForQt512.polkit-qt</li>
    <li>libsForQt514.breeze-icons</li>
    <li>libsForQt514.extra-cmake-modules</li>
    <li>libsForQt514.karchive</li>
    <li>libsForQt514.kauth</li>
    <li>libsForQt514.kcodecs</li>
    <li>libsForQt514.kcolorpicker</li>
    <li>libsForQt514.kcompletion</li>
    <li>libsForQt514.kconfig</li>
    <li>libsForQt514.kconfigwidgets</li>
    <li>libsForQt514.kcoreaddons</li>
    <li>libsForQt514.kcrash</li>
    <li>libsForQt514.kdbusaddons</li>
    <li>libsForQt514.kdoctools</li>
    <li>libsForQt514.kglobalaccel</li>
    <li>libsForQt514.kguiaddons</li>
    <li>libsForQt514.ki18n</li>
    <li>libsForQt514.kiconthemes</li>
    <li>libsForQt514.kimageannotator</li>
    <li>libsForQt514.kitemviews</li>
    <li>libsForQt514.kjobwidgets</li>
    <li>libsForQt514.kpackage</li>
    <li>libsForQt514.kservice</li>
    <li>libsForQt514.ktextwidgets</li>
    <li>libsForQt514.kwidgetsaddons</li>
    <li>libsForQt514.kwindowsystem</li>
    <li>libsForQt514.plasma-wayland-protocols</li>
    <li>libsForQt514.polkit-qt</li>
    <li>libsForQt514.sonnet</li>
    <li>lxqt.lxqt-build-tools</li>
    <li>qxmpp</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by veprbl** [comment]
I suppose this would also create wrappers for both existing wrappers and wrapped executables. Not sure what implications that would have. Another issue could be if file has executable bit set by accident.

Can we try to detect Mach-O and ELF in some other way? `stdenv` already ships with `isELF()` function.

File: pkgs/development/libraries/qt-5/5.12/default.nix [hunk 1]
```diff
@@ -121,7 +121,7 @@ let
       # to avoid cyclic dependencies between Qt modules.
       mkDerivation =
         import ../mkDerivation.nix
-        { inherit lib; inherit debug; wrapQtAppsHook = null; }
+        { inherit stdenv; inherit lib; inherit debug; wrapQtAppsHook = null; }
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.12/default.nix:124]:
```suggestion
        { inherit stdenv lib debug; wrapQtAppsHook = null; }
```

File: pkgs/development/libraries/qt-5/5.14/default.nix [hunk 2]
```diff
@@ -135,7 +135,7 @@ let
       # to avoid cyclic dependencies between Qt modules.
       mkDerivation =
         import ../mkDerivation.nix
-        { inherit lib; inherit debug; wrapQtAppsHook = null; }
+        { inherit stdenv; inherit lib; inherit debug; wrapQtAppsHook = null; }
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.14/default.nix:138]:
```suggestion
        { inherit stdenv lib debug; wrapQtAppsHook = null; }
```

File: pkgs/development/libraries/qt-5/5.14/default.nix [hunk 3]
```diff
@@ -147,7 +147,7 @@ let
 
       mkDerivationWith =
         import ../mkDerivation.nix
-        { inherit lib; inherit debug; inherit (self) wrapQtAppsHook; };
+        { inherit stdenv; inherit lib; inherit debug; inherit (self) wrapQtAppsHook; };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.14/default.nix:150]:
```suggestion
        { inherit stdenv lib debug; inherit (self) wrapQtAppsHook; };
```

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 4]
```diff
@@ -164,7 +164,7 @@ let
       # to avoid cyclic dependencies between Qt modules.
       mkDerivation =
         import ../mkDerivation.nix
-        { inherit lib; inherit debug; wrapQtAppsHook = null; }
+        { inherit stdenv; inherit lib; inherit debug; wrapQtAppsHook = null; }
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.15/default.nix:167]:
```suggestion
        { inherit stdenv lib debug; wrapQtAppsHook = null; }
```

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 5]
```diff
@@ -176,7 +176,7 @@ let
 
       mkDerivationWith =
         import ../mkDerivation.nix
-        { inherit lib; inherit debug; inherit (self) wrapQtAppsHook; };
+        { inherit stdenv; inherit lib; inherit debug; inherit (self) wrapQtAppsHook; };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.15/default.nix:179]:
```suggestion
        { inherit stdenv lib debug; inherit (self) wrapQtAppsHook; };
```

File: pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh [hunk 6]
```diff
@@ -85,7 +85,7 @@ wrapQtAppsHook() {
 
         find "$targetDir" ! -type d -executable -print0 | while IFS= read -r -d '' file
         do
-            patchelf --print-interpreter "$file" >/dev/null 2>&1 || continue
+            [ -x "$file" ] && [ ! -z "${qtWrapAllExecutables-}" ] || patchelf --print-interpreter "$file" >/dev/null 2>&1 || continue
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh:88]:
```suggestion
            [ -x "$file" ] && [ -n "${qtWrapAllExecutables-}" ] || patchelf --print-interpreter "$file" >/dev/null 2>&1 || continue
```

**Pull request closed without merging**